commit 06009f05c407b1b5333c9bec4fb390b440f36a16
Author: popcornmix <popcornmix@gmail.com>
Date:   Fri Apr 27 12:43:54 2012 +0100

    Update vchiq to match GPU version. Should still be compatible

Index: linux-3.2.46/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_arm.c
===================================================================
--- linux-3.2.46.orig/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_arm.c	2013-07-26 19:32:28.000000000 +0000
+++ linux-3.2.46/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_arm.c	2013-07-26 19:32:49.000000000 +0000
@@ -137,7 +137,7 @@
 ***************************************************************************/
 
 static inline USER_SERVICE_T *find_service_by_handle(
-	VCHIQ_INSTANCE_T instance, int handle )
+   VCHIQ_INSTANCE_T instance, int handle )
 {
    USER_SERVICE_T *user_service;
 
@@ -1094,7 +1094,7 @@
          char cr = '\n';
          if (copy_to_user(context->buf + context->actual - 1, &cr, 1))
          {
-	    context->actual = -EFAULT;
+            context->actual = -EFAULT;
          }
       }
    }
@@ -1383,11 +1383,11 @@
       }
       if(arm_state->use_notify_pending)
       {
-         send_pending = 1;
-         arm_state->use_notify_pending = 0;
+         send_pending = arm_state->use_notify_pending;
+         arm_state->use_notify_pending=0;
       }
       vcos_mutex_unlock(&arm_state->use_count_mutex);
-      if(send_pending)
+      while(send_pending--)
       {
          vcos_log_info( "%s sending VCHIQ_MSG_REMOTE_USE_ACTIVE", __func__);
          if ( vchiq_send_remote_use_active(state) != VCHIQ_SUCCESS)
@@ -1587,7 +1587,7 @@
       }
       if(!block_while_resume)
       {
-         arm_state->use_notify_pending = 1;
+         arm_state->use_notify_pending++;
          vcos_event_signal(&arm_state->hp_evt); /* hp task will check if we need to resume and also send use notify */
       }
 
@@ -1811,6 +1811,13 @@
    vcos_unused(state);
 }
 
+void vchiq_platform_conn_state_changed(VCHIQ_STATE_T *state, VCHIQ_CONNSTATE_T oldstate, VCHIQ_CONNSTATE_T newstate)
+{
+   vcos_unused(state);
+   vcos_unused(oldstate);
+   vcos_unused(oldstate);
+}
+
 
 /****************************************************************************
 *
Index: linux-3.2.46/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c
===================================================================
--- linux-3.2.46.orig/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c	2013-07-26 19:32:28.000000000 +0000
+++ linux-3.2.46/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c	2013-07-26 19:32:49.000000000 +0000
@@ -139,10 +139,12 @@
 static inline void
 vchiq_set_conn_state(VCHIQ_STATE_T *state, VCHIQ_CONNSTATE_T newstate)
 {
+   VCHIQ_CONNSTATE_T oldstate = state->conn_state;
    vcos_log_info("%d: %s->%s", state->id,
-      conn_state_names[state->conn_state],
+      conn_state_names[oldstate],
       conn_state_names[newstate]);
    state->conn_state = newstate;
+   vchiq_platform_conn_state_changed(state, oldstate, newstate);
 }
 
 static inline void
@@ -2686,15 +2688,30 @@
 
 VCHIQ_STATUS_T vchiq_send_remote_use(VCHIQ_STATE_T * state)
 {
-   return queue_message(state, NULL, VCHIQ_MAKE_MSG(VCHIQ_MSG_REMOTE_USE, 0, 0), NULL, 0, 0, 0);
+   VCHIQ_STATUS_T status = VCHIQ_RETRY;
+   if(state->conn_state != VCHIQ_CONNSTATE_DISCONNECTED)
+   {
+      status = queue_message(state, NULL, VCHIQ_MAKE_MSG(VCHIQ_MSG_REMOTE_USE, 0, 0), NULL, 0, 0, 0);
+   }
+   return status;
 }
 
 VCHIQ_STATUS_T vchiq_send_remote_release(VCHIQ_STATE_T * state)
 {
-   return queue_message(state, NULL, VCHIQ_MAKE_MSG(VCHIQ_MSG_REMOTE_RELEASE, 0, 0), NULL, 0, 0, 0);
+   VCHIQ_STATUS_T status = VCHIQ_RETRY;
+   if(state->conn_state != VCHIQ_CONNSTATE_DISCONNECTED)
+   {
+      status = queue_message(state, NULL, VCHIQ_MAKE_MSG(VCHIQ_MSG_REMOTE_RELEASE, 0, 0), NULL, 0, 0, 0);
+   }
+   return status;
 }
 
 VCHIQ_STATUS_T vchiq_send_remote_use_active(VCHIQ_STATE_T * state)
 {
-   return queue_message(state, NULL, VCHIQ_MAKE_MSG(VCHIQ_MSG_REMOTE_USE_ACTIVE, 0, 0), NULL, 0, 0, 0);
+   VCHIQ_STATUS_T status = VCHIQ_RETRY;
+   if(state->conn_state != VCHIQ_CONNSTATE_DISCONNECTED)
+   {
+      status = queue_message(state, NULL, VCHIQ_MAKE_MSG(VCHIQ_MSG_REMOTE_USE_ACTIVE, 0, 0), NULL, 0, 0, 0);
+   }
+   return status;
 }
Index: linux-3.2.46/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.h
===================================================================
--- linux-3.2.46.orig/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.h	2013-07-26 19:32:28.000000000 +0000
+++ linux-3.2.46/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.h	2013-07-26 19:32:49.000000000 +0000
@@ -500,6 +500,7 @@
 extern VCHIQ_STATUS_T
 vchiq_send_remote_use_active(VCHIQ_STATE_T * state);
 
-
+extern void
+vchiq_platform_conn_state_changed(VCHIQ_STATE_T* state, VCHIQ_CONNSTATE_T oldstate, VCHIQ_CONNSTATE_T newstate);
 
 #endif
Index: linux-3.2.46/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_ioctl.h
===================================================================
--- linux-3.2.46.orig/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_ioctl.h	2013-07-26 19:32:00.000000000 +0000
+++ linux-3.2.46/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_ioctl.h	2013-07-26 19:32:49.000000000 +0000
@@ -91,15 +91,15 @@
 #define VCHIQ_IOC_QUEUE_MESSAGE        _IOW(VCHIQ_IOC_MAGIC,  4, VCHIQ_QUEUE_MESSAGE_T)
 #define VCHIQ_IOC_QUEUE_BULK_TRANSMIT  _IOW(VCHIQ_IOC_MAGIC,  5, VCHIQ_QUEUE_BULK_TRANSFER_T)
 #define VCHIQ_IOC_QUEUE_BULK_RECEIVE   _IOW(VCHIQ_IOC_MAGIC,  6, VCHIQ_QUEUE_BULK_TRANSFER_T)
-#define VCHIQ_IOC_AWAIT_COMPLETION     _IOW(VCHIQ_IOC_MAGIC,  7, VCHIQ_AWAIT_COMPLETION_T)
-#define VCHIQ_IOC_DEQUEUE_MESSAGE      _IOW(VCHIQ_IOC_MAGIC,  8, VCHIQ_DEQUEUE_MESSAGE_T)
+#define VCHIQ_IOC_AWAIT_COMPLETION     _IOWR(VCHIQ_IOC_MAGIC, 7, VCHIQ_AWAIT_COMPLETION_T)
+#define VCHIQ_IOC_DEQUEUE_MESSAGE      _IOWR(VCHIQ_IOC_MAGIC, 8, VCHIQ_DEQUEUE_MESSAGE_T)
 #define VCHIQ_IOC_GET_CLIENT_ID        _IO(VCHIQ_IOC_MAGIC,   9)
-#define VCHIQ_IOC_GET_CONFIG           _IOW(VCHIQ_IOC_MAGIC, 10, VCHIQ_GET_CONFIG_T)
-#define VCHIQ_IOC_CLOSE_SERVICE        _IO(VCHIQ_IOC_MAGIC,  11)
-#define VCHIQ_IOC_USE_SERVICE          _IO(VCHIQ_IOC_MAGIC,  12)
-#define VCHIQ_IOC_RELEASE_SERVICE      _IO(VCHIQ_IOC_MAGIC,  13)
-#define VCHIQ_IOC_SET_SERVICE_OPTION   _IOW(VCHIQ_IOC_MAGIC, 14, VCHIQ_SET_SERVICE_OPTION_T)
-#define VCHIQ_IOC_DUMP_PHYS_MEM        _IOW(VCHIQ_IOC_MAGIC, 15, VCHIQ_DUMP_MEM_T)
+#define VCHIQ_IOC_GET_CONFIG           _IOWR(VCHIQ_IOC_MAGIC, 10, VCHIQ_GET_CONFIG_T)
+#define VCHIQ_IOC_CLOSE_SERVICE        _IO(VCHIQ_IOC_MAGIC,   11)
+#define VCHIQ_IOC_USE_SERVICE          _IO(VCHIQ_IOC_MAGIC,   12)
+#define VCHIQ_IOC_RELEASE_SERVICE      _IO(VCHIQ_IOC_MAGIC,   13)
+#define VCHIQ_IOC_SET_SERVICE_OPTION   _IOW(VCHIQ_IOC_MAGIC,  14, VCHIQ_SET_SERVICE_OPTION_T)
+#define VCHIQ_IOC_DUMP_PHYS_MEM        _IOW(VCHIQ_IOC_MAGIC,  15, VCHIQ_DUMP_MEM_T)
 #define VCHIQ_IOC_MAX                  15
 
 #endif
